include/wlr/types/wlr_surface.h

```diff
@@ -290,4 +298,23 @@ void wlr_surface_get_effective_damage(struct wlr_surface *surface,
 void wlr_surface_get_buffer_source_box(struct wlr_surface *surface,
 	struct wlr_fbox *box);
 
+/**
+ * Acquire a lock for the pending surface state.
+ *
+ * The state won't be committed before the caller releases the lock. Instead,
+ * the state becomes cached. The caller needs to use wlr_surface_unlock_cached
+ * to release the lock.
+ *
+ * Returns a surface commit sequence number for the cached state.
+ */
+uint32_t wlr_surface_lock_pending(struct wlr_surface *surface);
+
+/**
+ * Release a lock for a cached state.
+ *
+ * Callers should not assume that the cached state will immediately be
+ * committed. Another caller may still have an active lock.
+ */
+void wlr_surface_unlock_cached(struct wlr_surface *surface, uint32_t seq);
```

Comment:
Hmm, it's a little weird from a logical standpoint that `wlr_surface_lock_pending` is undone by `wlr_surface_unlock_cached`. I don't have an immediate idea for something better though.

Reply:
Yeah… Maybe `wlr_surface_cache_state` and `wlr_surface_uncache_state`?

Reply:
Hmm, yeah, idk if that's better to be honest. Leaving it up to you.